### PR TITLE
Web: Redo locking feature

### PR DESCRIPTION
**Pull request by kimlisa**
part of https://github.com/gravitational/teleport/issues/25168
enterprise PR: https://github.com/gravitational/teleport.e/pull/1237

recommend reviewing by commit (i plan to delete the old locks code after review)

it was hard to work with one size fit all data structure for the different types of resources especially when there are three types of pagination being used, so decided to re-do and have everything have its own table

design and ux is copied from access requests

in addition

- i didn't agree with supplying a manual add input box for every resource (especially when you can now use the search bar) so it only shows up for `logins`
- i copied the table display/formatting from other places, figured we can decide later what to customize if we want to
- add access checking, will not render lock tab if user doesn't have `lock.list` perm and does not render new lock screen if they don't have `lock.create and lock.update` perm
- add confirmation dialogue when user is deleting locks
- ** since we create a lock per resource select (this was how it worked before and figured it was okay), there's no guarantee that all will succeed in a row, so we allow user to re-attempt the failed ones
- supports all pagination types until we come up with another one ;;

Improvements for later:
- writing test and stories
- deduplicate code (shares similarities with access request)
- some cells are not aligned perfectly to center (you can tell on smaller screen with access requests)
- ** it looks like you can make batch lock request (as long the [target keys](https://github.com/gravitational/teleport/blob/fa54be823dc2681a498f042d5523b9f513eb30ee/api/proto/teleport/legacy/types/types.proto#L4309) are unique), like for example, locking one user and one device can be batched into one lock request. Locking two user has to be separate requests. Not sure if we wanted to support batching for less trip to back, but left it alone for another time.

#### Demo (using a smaller page size to demonstrate paginating):

https://user-images.githubusercontent.com/43280172/235037089-e415083a-fd0e-4385-babf-8d985a2726b5.mov

### CODE REVIEW

**Comment by kimlisa** [comment]
@rudream @avatus i would really appreciate it if ya'll can help me try to break this feature 🙏

File: web/packages/teleport/src/LocksV2/NewLock/ResourceList/common.tsx [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        <ButtonPrimary
+          width="134px"
+          size="small"
+          onClick={toggleResourceSelect}
+        >
+          Remove
+        </ButtonPrimary>
+      ) : (
+        <ButtonBorder width="134px" size="small" onClick={toggleResourceSelect}>
+          + Add to request
```
Comment (avatus) [inline web/packages/teleport/src/LocksV2/NewLock/ResourceList/common.tsx:86]: I don't think request is good language here but I'm not 100% sold on my suggestion. Best I could come up with. 
```suggestion
          + Add Target
```

If we use the language from the above comments like "Lock Request" then we can just keep this as is.

File: web/packages/teleport/src/LocksV2/NewLock/NewLock.tsx [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+          >
+            Clear Selections
+          </ButtonSecondary>
+        )}
+        <ButtonPrimary
+          width="182px"
+          onClick={() => setShowCheckout(true)}
+          disabled={!numAddedResources || isProcessing}
+        >
+          Proceed to Request
```
Comment (avatus) [inline web/packages/teleport/src/LocksV2/NewLock/NewLock.tsx:261]: Proceed to Lock? Proceed to Lock Request? Feel free to skip

File: web/packages/teleport/src/LocksV2/NewLock/NewLock.tsx [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+      alignItems="center"
+      justifyContent="space-between"
+      borderRadius={3}
+      p={3}
+      mt={5}
+      css={`
+        background: ${({ theme }) => theme.colors.spotBackground[0]};
+      `}
+    >
+      <Text bold>Resources Added ({numAddedResources})</Text>
```
Comment (avatus) [inline web/packages/teleport/src/LocksV2/NewLock/NewLock.tsx:244]:
```suggestion
      <Text bold>Targets Added ({numAddedResources})</Text>
```

File: web/packages/teleport/src/LocksV2/NewLock/LockCheckout/LockCheckout.tsx [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+          <Flex mb={3} alignItems="center">
+            <ArrowBack
+              fontSize={25}
+              mr={3}
+              onClick={onClose}
+              style={{ cursor: 'pointer' }}
+            />
+            <Box>
+              <Text typography="h4" color="text.primary" bold>
+                {locks.length} {pluralize(locks.length, 'Resource')} Selected
```
Comment (avatus) [inline web/packages/teleport/src/LocksV2/NewLock/LockCheckout/LockCheckout.tsx:208]:
```suggestion
                {locks.length} {pluralize(locks.length, 'Target')} Added
```

Again, not 100% sold here as you _are_ selecting resources TO lock so you can skip this one.

File: web/packages/teleport/src/LocksV2/NewLock/LockCheckout/LockCheckout.tsx [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+        to={
+          locks.length
+            ? {
+                pathname: cfg.getLocksRoute(),
+                state: { createdLocks: locks },
+              }
+            : cfg.getLocksRoute()
+        }
+      >
+        Back to Listings
```
Comment (avatus) [inline web/packages/teleport/src/LocksV2/NewLock/LockCheckout/LockCheckout.tsx:308]: Back to Locks?

File: web/packages/teleport/src/LocksV2/NewLock/ResourceList/Logins.tsx [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+    setAddedLogins([...addedLogins, { login: loginInput }]);
+  }
+
+  return (
+    <Box>
+      <Text mb={3}>
+        Listing logins are not supported. Use the input box below to manually
+        define a login to lock. <br />
+        Double check the spelling before adding.
```
Comment (avatus) [inline web/packages/teleport/src/LocksV2/NewLock/ResourceList/Logins.tsx:44]:
```suggestion
        Listing logins is not supported. Use the input box below to manually
        define a login to lock. <br />
        Double check the spelling before adding.
```

**Comment by zmb3** [comment]
> add access checking, will not render lock tab if user doesn't have lock.list perm and does not render new lock screen if they don't have lock.create and lock.update perm

We want to get away from this pattern of completely hiding parts of Teleport, because it prevents people from learning that these features exist.

For now it's fine, but we should start to prefer to always show the tabs and just indicate that the user doesn't have access.

File: web/packages/teleport/src/LocksV2/NewLock/NewLock.tsx [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+          <Flex alignItems="center">
+            <ArrowBack
+              as={Link}
+              fontSize={25}
+              mr={2}
+              title="Go Back"
+              to={cfg.getLocksRoute()}
+              style={{ cursor: 'pointer', textDecoration: 'none' }}
+            />
+            <Box>New Request</Box>
```
Comment (avatus) [inline web/packages/teleport/src/LocksV2/NewLock/NewLock.tsx:168]:
```suggestion
            <Box>New Lock Request</Box>
```

**Review by avatus — COMMENTED** [review]
Looks and feels really good to use now. I left some UI comments but more of a discussion rather than suggested changes. Feel free to push back/ignore them all.

File: web/packages/teleport/src/services/locks/types.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  expires: string;
+  createdAt: string;
+  createdBy: string;
+  targetLookup: Partial<Record<LockKind, string>>;
+  targets: LockTarget[];
+};
+
+export type LockTarget = {
+  kind: LockKind;
+  name: string;
```
Comment (zmb3) [inline web/packages/teleport/src/services/locks/types.ts:41]: This bit us in the past, right? It's not always a name.

Also, I'm fine with this as-is for simplicity, but note that the data model for the frontend doesn't match the backend now. On the backend, a lock target can have multiple different resources (so long as they are different types).

File: web/packages/teleport/src/services/locks/locks.ts [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+  },
+
+  deleteLock(id: string): Promise<void> {
+    return api.delete(cfg.getLocksUrlWithUuid(id));
+  },
+};
+
+export function makeLocks(json: any): Lock[] {
+  json = json || [];
+  return json.map(lock => makeLock(lock));
```
Comment (avatus) [inline web/packages/teleport/src/services/locks/locks.ts:38]:
```suggestion
  return json.map(makeLock);
```

File: web/packages/teleport/src/services/locks/types.ts (hunk 8 above)
Comment (kimlisa) [reply web/packages/teleport/src/services/locks/types.ts:41]: >This bit us in the past, right? It's not always a name.

i think this only applies to mapping backend with frontend fields, which isn't the case here. this type is only for frontend use. in the api response we get a mapping of targets, it wasn't very useful in its map form so I converted it into a list of LockTarget and added it as an extra field to the frontend model. i did it in this layer so we don't have to always process it when we use it in components.

we have a lot of frontend models that don't match the backend b/c we customize it further to suit our needs. i thought it was okay to do but maybe it shouldn't be done in the service layer. @ryanclark thoughts?

**Comment by kimlisa** [comment]
> > add access checking, will not render lock tab if user doesn't have lock.list perm and does not render new lock screen if they don't have lock.create and lock.update perm
> 
> We want to get away from this pattern of completely hiding parts of Teleport, because it prevents people from learning that these features exist.
> 
> For now it's fine, but we should start to prefer to always show the tabs and just indicate that the user doesn't have access.

is this something recently decided? i'll double check with xin in our upcoming monday meeting, but i thought we wanted it to be possible to turn off features? There are cases where we don't want to hide things, and thought those only applied when the user at least met the minimum criteria eg. most features look for at least a `list` access

File: web/packages/teleport/src/LocksV2/Locks/Locks.tsx [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+          <FeatureHeaderTitle>Session & Identity Locks</FeatureHeaderTitle>
+          <ButtonPrimary
+            as={NavLink}
+            to={cfg.getNewLocksRoute()}
+            ml="auto"
+            disabled={!canCreate}
+            title={
+              canCreate
+                ? ''
+                : 'You do not have access to create and update locks'
```
Comment (avatus) [inline web/packages/teleport/src/LocksV2/Locks/Locks.tsx:97]: Can we update locks?

**Review by avatus — APPROVED** [review]
Tested about everything I could think of. 
We can update some of the test plan points later but I plan on making a few large changes to that soon anyway.

I'd love to have @rudream comb over this as well

**Review by zmb3 — APPROVED** [review]
I was not able to break it. Thanks so much for this.

File: web/packages/teleport/src/LocksV2/Locks/Locks.tsx (hunk 10 above)
Comment (kimlisa) [reply web/packages/teleport/src/LocksV2/Locks/Locks.tsx:97]: you can https://github.com/gravitational/teleport/blob/860d91012c2a2087a58861a2e20682300b2eb7ea/lib/auth/auth_with_roles.go#L4876 (but no exposed api for the web to use atm), it's an `upsert` action that requires user to have both the create and update rule

**Pull request merged**